Fix providerID set to node name in idle allocations

rbtkess opened the pull request:
# Bug Report: ProviderID Set to Node Name Instead of Instance ID

## Summary

In idle allocations, the `providerID` property is incorrectly set to the Kubernetes node name (e.g., `ip-10-2-1-123.us-east-2.compute.internal`) instead of the cloud provider instance ID (e.g., `i-0abc123` for AWS EC2).

This makes it impossible to reconcile idle costs with cloud provider billing data, as billing systems use instance IDs, not Kubernetes node names.

## Expected Behavior

Idle allocations should have:
- `properties.node` set to Kubernetes node name (e.g., `ip-10-2-1-123.us-east-2.compute.internal`)
- `properties.providerID` set to cloud provider instance ID (e.g., `i-0abc123` for AWS, similar format for GCP/Azure)

## Actual Behavior

Idle allocations have:
- `properties.node` set to Kubernetes node name
- `properties.providerID` ALSO set to Kubernetes node name (incorrect)

Example:
```json
{
  "properties": {
    "cluster": "production",
    "node": "ip-10-2-1-123.us-east-2.compute.internal",
    "providerID": "ip-10-2-1-123.us-east-2.compute.internal"
  }
}
```

Should be:
```json
{
  "properties": {
    "cluster": "production",
    "node": "ip-10-2-1-123.us-east-2.compute.internal",
    "providerID": "i-0abc123"
  }
}
```

## Reproduction Steps

1. Deploy OpenCost to a cloud-based Kubernetes cluster (AWS, GCP, or Azure)
2. Query allocation API with idle by node:
   ```bash
   curl "http://your-opencost-endpoint:9003/allocation?window=1d&includeIdle=true&idleByNode=true"
   ```
3. Examine the `properties.providerID` field in idle allocations
4. Observe it contains the node name, not the cloud instance ID

## Root Cause

**File:** `pkg/costmodel/costmodel.go`
**Line:** 1799 (as of develop branch commit 78a8f1cf)
**URL:** https://github.com/opencost/opencost/blob/develop/pkg/costmodel/costmodel.go#L1799

The `ProviderID` field is set directly from `assetTotal.Node`:

```go
Properties: &opencost.AllocationProperties{
    Cluster:    assetTotal.Cluster,
    Node:       assetTotal.Node,
    ProviderID: assetTotal.Node,  // BUG: should be instance ID, not node name
},
```

### Why This Happens

The `AssetTotals` struct (defined in `core/pkg/opencost/totals.go:207`) does not have a `ProviderID` field:

```go
type AssetTotals struct {
    Start       time.Time
    End         time.Time
    Cluster     string
    Node        string      // Only has node NAME
    Count       int
    // ... no ProviderID field
}
```

When creating idle allocations, there's no `assetTotal.ProviderID` available, so the code uses `assetTotal.Node` as a fallback. However, this is incorrect - the actual instance ID needs to be looked up from the underlying asset data.

## Git Blame Analysis

This bug was introduced in commit `c8e9cf474` on 2023-03-22 during the PARC refactoring.

The issue is that the refactoring changed how idle allocations are created, but the new approach using `AssetTotals` lost the `ProviderID` information that was previously available.

## Impact

- Cannot reconcile idle costs with cloud provider billing data
- Billing reconciliation systems cannot match idle allocations to EC2/GCE/Azure VM instances
- Cost allocation strategies that distribute idle costs based on instance tags are broken
- Multi-cloud cost analysis tools cannot properly aggregate costs by provider instance

## Use Case Example

A common cost reconciliation workflow:
1. Fetch idle allocations from OpenCost (includes `providerID`)
2. Fetch billing data from AWS Cost Explorer (indexed by instance ID)
3. Join on `providerID` == instance ID to reconcile costs

Currently step 3 fails because `providerID` contains node names, not instance IDs.

## Environment

- OpenCost version: **v2.5.3 (latest stable release)** and develop branch (verified as of 2025-10-13)
- Kubernetes version: Any
- Cloud provider: AWS, GCP, Azure (any cloud provider)

**This bug affects all released versions since v1.102.0 (March 2023) when the PARC feature was introduced.**

## Proposed Fix

### Option 1: Add ProviderID to AssetTotals Struct

Modify `core/pkg/opencost/totals.go` to include `ProviderID`:

```go
type AssetTotals struct {
    Start       time.Time
    End         time.Time
    Cluster     string
    Node        string
    ProviderID  string  // Add this field
    Count       int
    // ...
}
```

Then modify `ComputeAssetTotals` to populate this field from asset data.

### Option 2: Look Up ProviderID from AssetSet

Modify `pkg/costmodel/costmodel.go:1798` to look up the providerID from the original `assetSet`:

```go
// Look up the actual providerID from assetSet
var providerID string
if assetTotal.Node != "" {
    // Find the node asset with this name
    for _, asset := range assetSet.Assets {
        if asset.Type() == opencost.AssetTypeNode {
            if nodeAsset, ok := asset.(*opencost.Node); ok {
                if nodeAsset.Properties.Name == assetTotal.Node {
                    providerID = nodeAsset.Properties.ProviderID
                    break
                }
            }
        }
    }
}

Properties: &opencost.AllocationProperties{
    Cluster:    assetTotal.Cluster,
    Node:       assetTotal.Node,
    ProviderID: providerID,  // Use actual instance ID
},
```

**Recommendation:** Option 1 is cleaner and more maintainable, but requires changes to the core data structures. Option 2 is a localized fix but involves iterating through assets on each idle allocation creation.



Fixes #3401

rbtkess committed:
Fix providerID set to node name in idle allocations

Idle allocations were setting providerID to the Kubernetes node name
instead of the cloud provider instance ID. This prevented reconciliation
with cloud billing data which uses instance IDs (e.g., AWS EC2 instance
IDs, GCP instance IDs, Azure VM IDs).

This commit:
1. Adds ProviderID field to AssetTotals struct
2. Populates ProviderID when computing asset totals
3. Uses ProviderID (not Node) when creating idle allocations

This enables proper reconciliation between OpenCost idle costs and
cloud provider billing data.

Fixes Issue #3401

Signed-off-by: Robert Kessler <rbtkess@users.noreply.github.com>
Signed-off-by: Robert Kessler <rob@cake.ai>

ameijer commented on pkg/costmodel/costmodel.go:
```diff
@@ -1796,7 +1796,7 @@ func computeIdleAllocations(allocSet *opencost.AllocationSet, assetSet *opencost
 			Properties: &opencost.AllocationProperties{
 				Cluster:    assetTotal.Cluster,
 				Node:       assetTotal.Node,
-				ProviderID: assetTotal.Node,
```
@rbtkess any way you could add some unit testing around this change? I think I agree with your change, just would like to reinforce that with testing

rbtkess replied:
Thank you @ameijer. Good call. Added unit tests in commit f498fc05:
 - `core/pkg/opencost/totals_providerid_test.go` - Tests AssetTotals.ProviderID population at node and cluster levels
  - `pkg/costmodel/idle_providerid_test.go` - Tests idle allocation ProviderID correctness for AWS, GCP, and Azure

 Tests cover:
  - ProviderID correctly set to cloud instance IDs (not node names)
  - Empty ProviderID handling for bare-metal/local clusters
  - Cluster-level aggregation scenarios
  - Multi-cloud environments

  All 8 tests pass.

rbtkess committed:
Add unit tests for idle allocation ProviderID fix

Add unit tests verifying that idle allocations correctly use cloud
provider instance IDs (from AssetTotals.ProviderID) rather than
Kubernetes node names.

Tests cover:
- AssetTotals.ProviderID population at node and cluster levels
- Idle allocation ProviderID correctness for AWS, GCP, and Azure
- Edge cases: empty ProviderID, cluster-level aggregation
- Multi-cloud and multi-node scenarios

Tests added:
- core/pkg/opencost/totals_providerid_test.go (4 tests)
- pkg/costmodel/idle_providerid_test.go (4 tests)

Addresses Issue #3401

Signed-off-by: Robert Kessler <rob@cake.ai>

ameijer reviewed (approved):
thank you for your contribution!
